File: resources/android/BiometricFunctions.kt
package com.iou.plugins.biometrics

import androidx.biometric.BiometricManager
import androidx.biometric.BiometricManager.Authenticators.BIOMETRIC_STRONG
import androidx.biometric.BiometricPrompt
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity
import com.nativephp.mobile.bridge.BridgeFunction
import com.nativephp.mobile.utils.NativeActionCoordinator
import org.json.JSONObject
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

/**
 * Namespace: "Biometric.*"
 */
object BiometricFunctions {

    private const val DEFAULT_EVENT_CLASS = "Native\\Mobile\\Events\\Biometric\\Completed"
    private const val AUTHENTICATE_TIMEOUT_SECONDS = 60L

    /**
     * Shows the system fingerprint/face prompt and blocks the calling (PHP) thread
     * until it reaches a terminal outcome.
     *
     * Every other interactive bridge function in this app's plugins (ContactsPicker,
     * DocumentPicker, Camera, Dialog) returns immediately and reports its result
     * later via a dispatched event — but `Native\Mobile\PendingBiometric::prompt()`
     * (nativephp/mobile core, unchanged) reads THIS call's own JSON return value
     * synchronously (`$decoded['status'] === 'success'`), not an event. Blocking
     * here is safe: per nativephp/mobile's own architecture notes, the native call
     * only blocks the PHP thread, never the UI thread, and — unlike an external
     * Camera activity — a BiometricPrompt is a dialog over the CURRENT activity,
     * so the app is never backgrounded while it's showing. The bounded timeout
     * below still guards against a stuck callback holding the PHP thread forever.
     */
    class Prompt(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val id = parameters["id"] as? String
            val eventClass = (parameters["event"] as? String) ?: DEFAULT_EVENT_CLASS

            val canAuthenticate = BiometricManager.from(activity).canAuthenticate(BIOMETRIC_STRONG)

            if (canAuthenticate != BiometricManager.BIOMETRIC_SUCCESS) {
                dispatchCompleted(eventClass, id, success = false)

                return mapOf("status" to "failed")
            }

            val latch = CountDownLatch(1)
            var success = false

            activity.runOnUiThread {
                val executor = ContextCompat.getMainExecutor(activity)
                val callback = object : BiometricPrompt.AuthenticationCallback() {
                    override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                        success = true
                        latch.countDown()
                    }

                    override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                        success = false
                        latch.countDown()
                    }

                    // Deliberately no override for onAuthenticationFailed: a single
                    // rejected fingerprint/face isn't terminal — the system prompt
                    // stays open for another attempt, so the latch must not release
                    // yet.
                }

                val label = activity.applicationInfo.loadLabel(activity.packageManager)

                val promptInfo = BiometricPrompt.PromptInfo.Builder()
                    .setTitle(label)
                    .setNegativeButtonText("Cancel")
                    .setAllowedAuthenticators(BIOMETRIC_STRONG)
                    .build()

                BiometricPrompt(activity, executor, callback).authenticate(promptInfo)
            }

            latch.await(AUTHENTICATE_TIMEOUT_SECONDS, TimeUnit.SECONDS)

            dispatchCompleted(eventClass, id, success)

            return mapOf("status" to if (success) "success" else "failed")
        }

        private fun dispatchCompleted(eventClass: String, id: String?, success: Boolean) {
            val payload = JSONObject().apply {
                put("success", success)
                if (id != null) {
                    put("id", id)
                }
            }

            NativeActionCoordinator.dispatchEvent(activity, eventClass, payload.toString())
        }
    }
}

File: resources/ios/BiometricFunctions.swift
import Foundation
import LocalAuthentication

private let biometricDefaultEventClass = "Native\\Mobile\\Events\\Biometric\\Completed"
private let biometricAuthenticateTimeoutSeconds: Double = 60
private let biometricPromptReason = "Use Face ID or Touch ID to continue."

/// Namespace: "Biometric.*"
enum BiometricFunctions {

    /// Shows the system Face ID/Touch ID prompt and blocks the calling (PHP) thread
    /// until it reaches a terminal outcome.
    ///
    /// Mirrors `BiometricFunctions.Prompt` on Android exactly: `LAContext.evaluatePolicy`
    /// is callback-based, so this wraps it in a `DispatchSemaphore` the same way Android
    /// wraps `BiometricPrompt`'s callback in a `CountDownLatch` — both exist because
    /// `Native\Mobile\PendingBiometric::prompt(): bool` (nativephp/mobile core, unchanged)
    /// reads THIS call's own JSON return value synchronously, not a later event. Blocking
    /// here is safe: per nativephp/mobile's own architecture notes, the native call only
    /// blocks the PHP thread, never the UI thread, on iOS exactly as on Android. The
    /// bounded timeout below still guards against a stuck callback holding the PHP thread
    /// forever.
    class Prompt: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            let id = parameters["id"] as? String
            let eventClass = (parameters["event"] as? String) ?? biometricDefaultEventClass

            let context = LAContext()

            // Empty string (not nil) hides the "Enter Password" fallback button
            // entirely, so this stays biometrics-only — matching Android's
            // BIOMETRIC_STRONG-only contract, with no device-credential fallback.
            context.localizedFallbackTitle = ""

            let policy = LAPolicy.deviceOwnerAuthenticationWithBiometrics

            var canEvaluateError: NSError?
            guard context.canEvaluatePolicy(policy, error: &canEvaluateError) else {
                dispatchCompleted(eventClass: eventClass, id: id, success: false)

                return BridgeResponse.success(data: ["status": "failed"])
            }

            let semaphore = DispatchSemaphore(value: 0)
            var success = false

            context.evaluatePolicy(policy, localizedReason: biometricPromptReason) { result, _ in
                success = result
                semaphore.signal()
            }

            _ = semaphore.wait(timeout: .now() + biometricAuthenticateTimeoutSeconds)

            dispatchCompleted(eventClass: eventClass, id: id, success: success)

            return BridgeResponse.success(data: ["status": success ? "success" : "failed"])
        }

        private func dispatchCompleted(eventClass: String, id: String?, success: Bool) {
            var payload: [String: Any] = ["success": success]
            if let id = id {
                payload["id"] = id
            }

            DispatchQueue.main.async {
                LaravelBridge.shared.send?(eventClass, payload)
            }
        }
    }
}
